crates/turbopack-node/src/render/render_proxy.rs

```diff
@@ -75,13 +76,11 @@ async fn run_proxy_operation(
         .send(RenderProxyOutgoingMessage::Headers { data: &data })
         .await?;
 
-    let body = body.await?;
+    let mut body = body.await?.read();
```

Comment:
When `StreamRead` is merged into `Stream`, you would `fork`/`clone` the cached stream to consume it, while leaving the cached stream untouched.

```suggestion
    let mut body = body.await?.clone();
```

Reply:
> The index can't be used on the original value, because it would only screw up all readers if it were ever read from originally.

Rust mutability prevents reading from the cached value since you don't have a mutable reference to it.

Also readers won't references the original stream. They have a cloned stream with a shared `StreamState`. StreamState is still immutable in that sense since we never remove items from `pulled`.

With

``` rs
pub struct Stream<T> {
	    inner: Arc<Mutex<StreamState<T>>>,
	    /// Number of items consumed in the current segment.
	    /// Might revert back to 0 when reached the end of a segment
	    consumed: usize,
}
```

you would need a `&mut Stream` to read from it, which is only possible if you are the only owner.

Reply:
> Rust mutability prevents reading from the cached value since you don't have a mutable reference to it.

What I mean is imagine you clone from an already mutated stream:

```rust
let mut s = Stream(…);

// This advances `s.index`
s.next.await?;

// Inherits the advanced index
s.clone()
```

So by mutating the index, any clones that want to get access to the full stream would be broken. I don't see why we want to store this on the struct itself anyways, we don't store iteration index as part of a `Vec<T>`. If you want to read and iterate, you alloc your own `Iter` to hold your position.